alpa/torch/nn/__init__.py

```diff
@@ -0,0 +1,490 @@
+        # Temporarily load the state back onto self.stateless_model
+        old_state = _swap_state(self.stateless_model, self.all_names_map,
+                                list(params.values()) + list(buffers.values()))
+        try:
+            return buffers, self.stateless_model(*args, **kwargs)
+        finally:
+            # Remove the loaded state on self.stateless_model
+            _swap_state(self.stateless_model, self.all_names_map, old_state)
+
+
+def functionalize(module: torch.nn.Module, inputs, batch_dim,
+                  num_micro_batches):
+    """Returns:
+        - `module_func`: a function that has same logic as x.forward but
+        callable with either PT or Alpa inputs. It:
+            - wraps the original inputs in a tuple
+            - takes `params` and `bufs` as extra at beginning of input list
+            - produces `bufs` as extra output at beginning of output list
+            - all calls are made compatible with Alpa, e.g.:
+                - replaces all unexpandable module calls (e.g. nn.Conv2d) with
+                  equivalent `torch.*` function calls
+                - replaces all torch.nn.functional calls that has in-place ops
+                  (e.g. F.batch_norm) with equivalent `atorch.*` function calls
+                  that has buffer as part of output
+                - complex torch function calls (e.g. F.dropout) are decomposed
+                  and implemented with `torch.*` calls
+        - `params`: a dict of shape-only tensors representing the trainable
+           parameters of the module.
+           In PT format if "local", in Alpa format if "dist".
+        - `bufs`: a dict of shape-only tensors representing the no-gradient
+           parameters of the module.
+           In PT format if "local", in Alpa format if "dist".
+    Throws error if x.forward:
+        - has in-place ops
+        - or, has data-dependent control flow
+        - or, has other graph-breaking statements (e.g. `print()`) that
+          prevents the program from being captured as a single graph
+          (only in "dist" mode)
+    """
+
+    # This param/buffer name map is used for mapping from FQN in original
+    # PyTorch model to FQN in PyTorch FX IR.
+    tensor_to_name_map = {}
+
+    all_tensors_pt_orig = dict(named_parameters(module))
+    all_tensors_pt_orig.update(dict(named_buffers(module)))
+
+    for k, v in all_tensors_pt_orig.items():
+        assert v not in tensor_to_name_map
+        tensor_to_name_map[v] = {"orig_name": k}
+
+    def add_transformed_name(tensor_to_name_map, k, v):
+        assert v in tensor_to_name_map
+        assert "transformed_name" not in tensor_to_name_map[v]
+        tensor_to_name_map[v]["transformed_name"] = k
+
+    if atorch.mode() == "dist":
+        # In dist mode, use TorchDynamo to enforce:
+        # 1) no data-dependent control flow
+        # 2) no graph break points
+        # 3) no in-place ops
+
+        def convert_pt_module_to_jax_func(module):
```

Comment:
Should be `convert_pt_module_to_alpa_func`. The return function still includes pytorch calls.